docs/proposals/api-types-package-structure.md

```diff
@@ -0,0 +1,518 @@
+│       │   ├── defaulting
+│       │   │   ├── generated_defaulters.go
+│       │   │   └── generated_default_stubs.go
+│       │   ├── types.go
+│       │   └── validation
+│       │       ├── generated_validation.go
+│       │       └── manual_validation_overrides.go
+│       └── resource3
+│           ├── conversion
+│           │   ├── generated_conversion.go
+│           │   └── manual_conversion_overrides.go
+│           ├── defaulting
+│           │   ├── generated_defaulters.go
+│           │   └── generated_default_stubs.go
+│           ├── types.go
+│           └── validation
+│               ├── generated_validation.go
+│               └── manual_validation_overrides.go
+└── unversioned
+    ├── meta
+    │   └── meta.go
+    ├── resource
+    │   └── quantity.go
+    ├── time
+    │   └── time.go
+    └── types.go
+```
+
+### Package: `apis/`
+
+Multiple groups may live in the same tree. Each group gets its own directory
+under `apis/`. We'll assume an `apis/` directory of this structure in all tools.
+In the above example, we show one group, named 'group'.
+
+TBD: Either test code will live here, or no code.
+
+### Package: `apis/unversioned/`
+
+This package and sub-packages are allowed to be imported and used by any
+group/version/resource. There is no internal/versioned distinction. Types in
+this directory in the main kubernetes repository are reusable by anyone in any
+API; creators of 3rd party APIs should think carefully before creating their own
+unversioned package.
+
+### Package: `apis/group/`
+
+Multiple versions live in a group. Each version gets its own subdirectory.
+
+Additionally, a directory for each resource has a types.go file containing an
+"internal" version of the group's objects. These files are used as a
+destination/source for conversion functions.
+
+### Package: `apis/group/install`
+
+This package installs *all* of the versions of the group, meaning it imports
+each nested install/ package.
+
+### Package: `apis/group/v1`
+
+'v1' is taken to be the current stable version of this api.
+
+Multiple resource directories (we show 'resource1' and 'resource2') each have a
+`types.go`, containing the type definitions for this group/version/resource.
+
+This package has tightly controlled imports. It's allowed to include the
+Kubernetes unversioned API types (to be moved in/under `pkg/apis/unversioned`)
+and little else. In particular, it *must not* import its parent `apis/group`.
+
+### Package: `apis/group/v1/install`
+
+This package, if imported, registers (via an `init()`) the v1 version of every
+resource in this group, its conversion, validation, and defaulting functions.
+
+We have separate install packages for every version to allow deliberate import
+choices to be made.
+
+### Package: `apis/group/v1/resource1/conversion`
+
+This package contains conversion functions, and exports a Register() function
+which will register them (but does not register them as a side effect of being
+imported).
```

Comment:
nevermind, it's late and I apparently can't read... install calls the Register() functions